Document Stage tagging (#649)
grafana/k6-docs: src/data/markdown/translated-guides/en/02 Using k6/08 Tags and Groups.md

- *System tags* are tags that k6 automatically assigns.
- *User-defined* tags are tags that you add when you write your script.

> ## System tags

Reviewer: What was the reasoning behind bringing the system tags section as the same level as tags? 😸

Author: The reason was for creating the right bar with all of this section with a specific link. AFAIK, only the h2 headers are added to the list.

![Screenshot from 2022-05-02 12-10-41](https://user-images.githubusercontent.com/2103732/166218657-3a5e143c-539e-48c0-bafb-65af9acc35d1.png)

Reviewer: Makes sense. Thanks for the clarification ❤️

| `ip`          | The IP address of the remote server                                                                                               |
| `ocsp_status` | the [Online Certificate Status Protocol (OCSP)](/using-k6/protocols/ssl-tls/online-certificate-status-protocol-ocsp) HTTPS status |

> ## User-defined tags

Reviewer: Same question about making this a `##` 👀

</CodeGroup>

> ## Test-wide tags

Reviewer: Same question having this as a `##` 👀

Besides attaching tags to requests, checks, and custom metrics, you can set test-wide tags across all metrics.
There are two ways you can set these tags:

- In the CLI, using one or more `--tag NAME=VALUE` flags

- In the script itself:

  <CodeGroup labels={["test-wide-tags.js"]} lineNumbers={[true]}>

  ```javascript
  export const options = {
    tags: {
      name: 'value',
    },
  };
  ```

  </CodeGroup>

## Code-defined tags

In the case, a user-defined tag with advanced logic for handling which tag to set is required then it's possible doing it by defining the tag from the code.

> It's possible to use the [k6/execution.vu.tags](/javascript-api/k6-execution/#vu) API for assigning an object property with the name and the value of the required tag. Like the following example for tracking the container group from the nested groups, it would be helpful for aggregating sub-metrics for nested groups.

Reviewer: ```suggestion
To support advanced tagging logics and workflows, it is also possible to directly manipulate them from scripts' code.

[k6/execution.vu.tags](/javascript-api/k6-execution/#vu) object's properties can indeed be directly assigned new key/value pairs to define new tags dynamically. This can prove useful, as demonstrated in the following example, to track a container's group from nested groups, and aggregating nested group's sub-metrics.
```

Here's a proposal I'm not too happy with, as I was not 100% sure I understand what you meant in detail. Please don't hesitate to adjust accordingly. On a sidenote: I encountered the issue recently, and I don't believe we even document sub-metrics publicly?

```javascript
import http from 'k6/http'
import exec from 'k6/execution'
import { group } from 'k6'

export const options = {
    thresholds: {
        'http_reqs{container_group:main}': ['count==3'],
        'http_req_duration{container_group:main}': ['max<1000'],
    },
};

export default function () {
    exec.vu.tags['container_group'] = 'main'
    
    group('main', function () {
        http.get('https://test.k6.io')
        group('sub', function () {
            http.get('https://httpbin.test.k6.io/anything')
        });
        http.get('https://test-api.k6.io')
    });

    delete exec.vu.tags['container_group']

    http.get('https://httpbin.test.k6.io/delay/3')
}
```

> Using the same API is even possible to get any already set user-defined and/or system-defined tag:

Reviewer: ```suggestion
Using the same API, it is also possible to retrieve any already set user-defined and/or system-defined tag:
```

```javascript
import exec from 'k6/execution'

export default function() {
    let tag = exec.vu.tags['scenario']
    console.log(tag) // default
} 
```

## Tagging stages

> Thanks to some defined helper functions in the [k6-jslib-utils](/javascript-api/jslib/utils) project, it's possible to add a tag with the current ongoing stage for the executors that supports the `stages` option.

Reviewer: ```suggestion
Thanks to some of the helper functions defined in the [k6-jslib-utils](/javascript-api/jslib/utils) project, it is possible to tag executors supporting the `stages` option with the current ongoing stage.
```

```javascript
import http from 'k6/http';
import exec from 'k6/execution';
import { group } from 'k6';

export const options = {
    thresholds: {
        'http_reqs{container_group:main}': ['count==3'],
        'http_req_duration{container_group:main}': ['max<1000'],
    },
};

export default function () {
    exec.vu.tags['container_group'] = 'main';
    
    group('main', function () {
        http.get('https://test.k6.io');
        group('sub', function () {
            http.get('https://httpbin.test.k6.io/anything');
        });
        http.get('https://test-api.k6.io');
    });

    delete exec.vu.tags['container_group'];

    http.get('https://httpbin.test.k6.io/delay/3');
}
```

Using the same API, it is also possible to retrieve any already set user-defined and/or system-defined tag:

```javascript
import exec from 'k6/execution'

export default function() {
    let tag = exec.vu.tags['scenario'];
    console.log(tag); // default
} 
```

## Tagging stages

> Thanks to some of the helper functions defined in the [k6-jslib-utils](/javascript-api/jslib/utils) project, it is possible to tag executors supporting the `stages` option with the current ongoing stage.

Author: ```suggestion
Thanks to some helper functions in the [k6-jslib-utils](/javascript-api/jslib/utils) project, if an executor supports the `stages` option, you can add a tag with the current ongoing stage. The tag will be added to all the samples collected during the iteration.
```
